README.md
peritus-tc-tools
================

A few OpenTSDB tcollector tools and plugins

[tc-ping][1]
------------
High performance replacement for smokeping, collecting latency and loss data using [fping][6].

![](https://raw.github.com/PeritusConsulting/peritus-tc-tools/master/tc-ping/tc-ping-details-one-host.png)

Metrics:

* ```icmp.lossrate``` - percentage
* ```icmp.latency.avg``` - milliseconds
* ```icmp.latency.max``` - milliseconds

Tags:

* ```dsthost```
* ```host```
* ```tos```

[tc-dhcpd-pools][2]
-------------------
High performance collection of ISC dhcpd pool lease statistics using [dhcpd-pools][7].
![](https://raw.github.com/PeritusConsulting/peritus-tc-tools/master/tc-dhcpd-pools/tc-dhcpd-pools-graph-screenshot.png)

Metrics:

* ```dhcpd.pool.size```
* ```dhcpd.pool.usage.absolute```
* ```dhcpd.pool.usage.percent```
* ```dhcpd.pool.inactive```
* ```dhcpd.pool.backup```

Tags:

* ```host```
* ```dhcppoolname```

[tc-isc-dhcpd-log-stats][3]
---------------------------
Simple log "parser" for ISC dhcpd collecting DHCP event statistics (DISCOVER, OFFER, ACK, REQUEST, etc)
![](https://raw.github.com/PeritusConsulting/peritus-tc-tools/master/tc-isc-dhcpd-log-stats/tc-isc-dhcpd-log-stats-graph-request-ack.png)

Metrics:

* ```dhcp.events.discover```
* ```dhcp.events.offer```
* ```dhcp.events.request```
* ```dhcp.events.ack```
* ```dhcp.events.nak```
* ```dhcp.events.release```
* ```dhcp.events.inform```
* ```dhcp.events.addforwardmap```
* ```dhcp.events.removeforwardmap```
* ```dhcp.errors.noaddforwardmap```
* ```dhcp.errors.noaddreversemap```
* ```dhcp.errors.notsec```
* ```dhcp.errors.bindupdatexidmismatch```
* ```dhcp.errors.uidduplicate```
* ```dhcp.errors.bindupdaterejected```
* ```dhcp.errors.others```
* ```dhcp.total```
* ```tcollector.collectors.isc-dhcpd-log-stats.elapsedtime```

Tags:

* ```host```
* ```instance```


[tc-atmail][4]
--------------
Simple script counts [Atmail][8] mysql database rows.
![](https://raw.github.com/PeritusConsulting/peritus-tc-tools/master/tc-atmail/tc-atmail-graph-screenshot.png)

Metrics:

* ```atmail.adminusers```
* ```atmail.domains```
* ```atmail.mailrelays```
* ```atmail.sharedfiles```
* ```atmail.users```
* ```atmail.usersessions```
* ```atmail.log.virus```
* ```atmail.log.spam```
* ```atmail.log.sentmail```
* ```atmail.log.recvmail```
* ```atmail.log.logins```
* ```atmail.log.errors```
* ```tcollector.collectors.atmail.elapsedtime```

Tags:

* ```host```

[tc-opensips][5]
----------------
Simple script counts [opensips][9] mysql database rows.
![](https://raw.github.com/PeritusConsulting/peritus-tc-tools/master/tc-opensips/tc-opensips-graph-locations-subscribers.png)

Metrics:

* ```opensips.subscribers```
* ```opensips.locations```
* ```tcollector.collectors.opensips.elapsedtime```

Tags:

* ```host```


  [1]: https://github.com/PeritusConsulting/peritus-tc-tools/tree/master/tc-ping
  [2]: https://github.com/PeritusConsulting/peritus-tc-tools/tree/master/tc-dhcpd-pools
  [3]: https://github.com/PeritusConsulting/peritus-tc-tools/tree/master/tc-isc-dhcpd-log-stats
  [4]: https://github.com/PeritusConsulting/peritus-tc-tools/tree/master/tc-atmail
  [5]: https://github.com/PeritusConsulting/peritus-tc-tools/tree/master/tc-opensips
  [6]: http://fping.org/
  [7]: http://dhcpd-pools.sourceforge.net/
  [8]: http://atmail.com/
  [9]: http://www.opensips.org/


tc-ping/CHANGELOG.md
tc-ping.py changelog
---------------------------
2014-08-23 - v1.1.2 - Changed default interface tag from nothing to "default" for easier viewing in OpenTSDB.

2014-08-23 - v1.1.1 - Added interface option. Useful for comparing wired and wireless, for example.

2014-04-06 - v1.1.0 - Moved into peritus-tc-tools repo and renamed from tcollectorping to tc-ping

2014-03-10 - v1.1.0 - Added ToS flag to fping for setting QoS priorities on pings. 
                      Also tags outgoing metrics with ToS value for easy comparison.

2014-03-09 - v1.0.0 - Initial release


tc-ping/tc-ping.py
#!/usr/bin/env python

"""

peritus-tc-tools/tc-ping/tc-ping.py - v 1.1.2
OpenTSDB tcollector wrapper for fping3

2014-08-23 Stian Ovrevage <stian@peritusconsulting.no>

Pings hosts every 3 seconds, outputs statistic every 10 seconds
which are formatted into OpenTSDB tcollector compatible format.
Metrics gathered:
  icmp.latency.avg
  icmp.latency.max
  icmp.latency.lossrate

For more information on fping:
http://fping.org/fping.1.html

fping output format:
  up.com : xmt/rcv/%loss = 1/1/0%, min/avg/max = 0.98/0.98/0.98
  down.org : xmt/rcv/%loss = 1/0/100%

"""

import subprocess
import time
import sys
import os.path
import socket

''' Text file containing hostnames or IPs to be pinged, one per line '''
hostfile = "/opt/tcollector/collectors/etc/hosts.txt"

'''
Interval between pings to an individual host, in milliseconds
Recommended 3000. 3 seconds.
'''
pinginterval = "3000"

'''
Interval between printing statistics, and hence, sending to OpenTSDB
Recommended 10 seconds.
'''
statsinterval = "10"

'''
Number of pings to send in one session. tcollector will re-start 
this script when it finishes and it will then re-read the host-list
Recommended 1200 pings. Will run each session 
for approx 1 hour before respawning.
'''
pingcount = "1200"

'''
Type-of-service byte. Used for QoS marking of packets. Good reference
and mapping og TOS-to-DSCP here: http://www.tucny.com/Home/dscp-tos
Default should be 0. DSCP cs4 (32) is TOS 128. DSCP ef (46) is TOS 184.
'''
tos = "0"

'''
Interface. Usefull for pinging hosts via different interfaces, such as wired
and wireless. Have a different tc-ping.py for each interface-value.
Default is empty for auto. Example: interface = "wlan0"
'''
interface = ""

''' host tag for metrics. Use system fqdn hostname by default '''
host = socket.getfqdn()

''' create interface tag string, if interface is defined '''
if len(interface) > 0:
  interface_tag = "interface=" + interface
else:
  interface_tag = "interface=default"


command = ['fping -I "' + interface + '" -B 1 -C ' + pingcount + ' -D -r0 -O ' + tos + ' -Q '
           + statsinterval + ' -p ' + pinginterval + ' < ' + hostfile]

def main():
    if not os.path.isfile(hostfile):
        print >> sys.stderr, 'error: '+hostfile+' does not exist'
        return 1

    try:
        p = subprocess.Popen(command, shell=True, stderr=subprocess.PIPE, bufsize=0)
        line = p.stderr.readline()
        while line:
            ts = int(time.time())

            line = line.replace(':','')
            line = line.replace('=','')
            line = line.replace(',','')
            line = line.replace('%','')
            line = line.replace('/',' ')
            
            fields = line.split()
            
            # Only process lines including a "xmt", this excludes time lines etc
            if 'xmt' in fields:
                print("icmp.lossrate " + str(ts) + " " + fields[6] +
                      " host=" + host + " dsthost=" + fields[0] + " tos=" + tos + " " + interface_tag)
                # RTT numbers only included if at least one packet received:
                if 'avg' in fields:
                    print("icmp.latency.avg " + str(ts) + " " + fields[11] +
                          " host=" + host + " dsthost=" + fields[0] + " tos=" + tos + " " + interface_tag)
                    print("icmp.latency.max " + str(ts) + " " + fields[12] +
                          " host=" + host + " dsthost=" + fields[0] + " tos=" + tos + " " + interface_tag)

            line = p.stderr.readline()
        
    except OSError, err:
        print >> sys.stderr, 'Got execption running command "%s": %s' % (command, err)

if __name__ == "__main__":
    main()


tc-wlan/tc-wlan.py
#!/usr/bin/env python

"""

peritus-tc-tools/tc-wlan/tc-wlan.py - v 1.0.0
OpenTSDB tcollector wrapper for iwlist scan

2014-08-23 Stian Ovrevage <stian@peritusconsulting.no>

Metrics gathered:
  wlan.channel
  wlan.bitrate
  wlan.quality_level
  wlan.signal_level
  wlan.adjecent_networks

iwlist scan output format:
          Cell 01 - Address: 00:31:96:41:8C:8B
                    ESSID:"SomeSSIDName"
                    Protocol:IEEE 802.11bg
                    Mode:Master
                    Frequency:2.462 GHz (Channel 11)
                    Encryption key:on
                    Bit Rates:54 Mb/s
                    Extra:wpa_ie=dd160050f201010
                    IE: WPA Version 1
                        Group Cipher : TKIP
                        Pairwise Ciphers (1) : TKIP
                        Authentication Suites (1) : PSK
                    Quality=100/100  Signal level=85/100

Tested with:
D-link DWA-121 - Recommended. High update rate. Apparently high precision.
Edimax EW-7811Un - Only updates values every 2-3 minutes
Asus USB-N10 - Only updates values every 2-3 minutes

"""

import subprocess
import time
import sys
import os.path
import socket
import re

'''
How long between scans, in seconds. Note that an interval of 10 will not give a result
every 10 seconds since iwlist typically takes a second or two to send the scan
command to the NIC and receive an answer. Recommended: 10. Certain NICs can scan
very often and 3 second interval has been used successfully. But this could lead
to overheating over time, so use caution with very short intervals.
'''
scan_interval = 10


'''
Which interface to use.
'''
interface = "wlan0"

''' host tag for metrics. Use system fqdn hostname by default '''
host = socket.getfqdn()


command = ['iwlist ' + interface + ' scan']

def main():

    ssid = ""
    re_ssid = re.compile(r'\s*ESSID:\"([^\n\r"]+)\"\s*')
    channel = ""
    re_channel = re.compile(r'\s*Frequency:.*Channel ([0-9]+)\s*')
    bitrate = ""
    re_bitrate = re.compile(r'\s*Bit Rates:([0-9]+) Mb/s\s*')
    quality_level = ""
    signal_level = ""
    re_levels = re.compile(r'\s*Quality=([0-9]+)/100.*Signal level=([0-9]+)/100\s*')
    re_signal_level = re.compile(r'\s*Signal level=([0-9]+)/100\s*')

    re_cell = re.compile(r'\s*Cell ([0-9]+) - Address:.*')

    adjecent_networks = 0

    try:
        while(True):
            ts = int(time.time())

            p = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, bufsize=0)
            line = p.stdout.readline()
            while line:
                if re_cell.match(line):
                    adjecent_networks = adjecent_networks + 1
                    ssid, channel, bitrate, quality_level, signal_level = "", "", "", "", ""

                if re_ssid.match(line):
                    ssid = re_ssid.match(line).groups()[0]

                if re_channel.match(line):
                    channel = re_channel.match(line).groups()[0]
                    print("wlan.channel " + str(ts) + " " + channel + " host=" + host + " interface=" + interface + " ssid=" + ssid)

                if re_bitrate.match(line):
                    bitrate = re_bitrate.match(line).groups()[0]
                    print("wlan.bitrate " + str(ts) + " " + bitrate + " host=" + host + " interface=" + interface + " ssid=" + ssid)

                if re_levels.match(line):
                    quality_level = re_levels.match(line).groups()[0]
                    signal_level = re_levels.match(line).groups()[1]
                    print("wlan.quality_level " + str(ts) + " " +  quality_level  + " host=" + host + " interface=" + interface + " ssid=" + ssid)
                    print("wlan.signal_level " + str(ts) + " " +  signal_level + " host=" + host + " interface=" + interface + " ssid=" + ssid)

                if re_signal_level.match(line):
                    signal_level = re_signal_level.match(line).groups()[0]
                    print("wlan.signal_level " + str(ts) + " " + signal_level + " host=" + host + " interface=" + interface + " ssid=" + ssid)

                line = p.stdout.readline()

            ''' Output statistic on number of adjecent networks '''
            print("wlan.adjecent_networks " + str(ts) + " " +  str(adjecent_networks) + " host=" + host + " interface=" + interface + " ssid=" + ssid)
            adjecent_networks = 0

            time.sleep(scan_interval)

    except OSError, err:
        print >> sys.stderr, 'Got execption running command "%s": %s' % (command, err)

if __name__ == "__main__":
    main()


tc-atmail/tc-atmail.sh
#!/bin/sh
# 
# peritus-tc-tools/tc-atmail/tc-atmail.sh
# Count some tables from @mail MySQL database 
# and output it to OpenTSDB tcollector
# 
# 2014-04-06 Stian Øvrevåge <stian@peritusconsulting.no>
#
# This script will simply count the number of rows for some of the atmail
# tables and output the metrics in tcollector compatible format.
#
# It will also output a metric of its own running time in microseconds in
# order to evaluate performance over time as counting mysql-rows on huge
# databases might result in performance degradation.
#
# Instructions:
#   Put this file into a tcollector/collectors/XX/ folder where XX is how
#   often you want the script to run. 300 should be reasonable and gives
#   statistics every 5 minutes. Make sure the script is runnable.  
# 
#   You can run the script on the @mail server itself or another host.
#   Make sure you create a user with permissions to connect if using a 
#   remote host. And update the config below.
#
#   You can run the script itself on the command-line to verify that 
#   it is working properly.
#

MYSQL_USER=
MYSQL_PASS=
MYSQL_HOST=localhost
MYSQL_DB=atmail

TAG_HOST=`hostname`

STARTTIME=$(date +%s%N)

echo atmail.adminusers `date +%s` `echo "USE $MYSQL_DB; SELECT count(*) FROM AdminUsers;" | mysql --host $MYSQL_HOST --user $MYSQL_USER -p$MYSQL_PASS --skip-column-names` host=$TAG_HOST
echo atmail.domains `date +%s` `echo "USE $MYSQL_DB; SELECT count(*) FROM Domains;" | mysql --host $MYSQL_HOST --user $MYSQL_USER -p$MYSQL_PASS --skip-column-names` host=$TAG_HOST
echo atmail.mailrelays `date +%s` `echo "USE $MYSQL_DB; SELECT count(*) FROM MailRelay;" | mysql --host $MYSQL_HOST --user $MYSQL_USER -p$MYSQL_PASS --skip-column-names` host=$TAG_HOST
echo atmail.sharedfiles `date +%s` `echo "USE $MYSQL_DB; SELECT count(*) FROM SharedFiles;" | mysql --host $MYSQL_HOST --user $MYSQL_USER -p$MYSQL_PASS --skip-column-names` host=$TAG_HOST
echo atmail.users `date +%s` `echo "USE $MYSQL_DB; SELECT count(*) FROM Users;" | mysql --host $MYSQL_HOST --user $MYSQL_USER -p$MYSQL_PASS --skip-column-names` host=$TAG_HOST
echo atmail.usersessions `date +%s` `echo "USE $MYSQL_DB; SELECT count(*) FROM UserSession;" | mysql --host $MYSQL_HOST --user $MYSQL_USER -p$MYSQL_PASS --skip-column-names` host=$TAG_HOST
echo atmail.log.virus `date +%s` `echo "USE $MYSQL_DB; SELECT count(*) FROM Log_Virus;" | mysql --host $MYSQL_HOST --user $MYSQL_USER -p$MYSQL_PASS --skip-column-names` host=$TAG_HOST
echo atmail.log.spam `date +%s` `echo "USE $MYSQL_DB; SELECT count(*) FROM Log_Spam;" | mysql --host $MYSQL_HOST --user $MYSQL_USER -p$MYSQL_PASS --skip-column-names` host=$TAG_HOST
echo atmail.log.sentmail `date +%s` `echo "USE $MYSQL_DB; SELECT count(*) FROM Log_SendMail;" | mysql --host $MYSQL_HOST --user $MYSQL_USER -p$MYSQL_PASS --skip-column-names` host=$TAG_HOST
echo atmail.log.recvmail `date +%s` `echo "USE $MYSQL_DB; SELECT count(*) FROM Log_RecvMail;" | mysql --host $MYSQL_HOST --user $MYSQL_USER -p$MYSQL_PASS --skip-column-names` host=$TAG_HOST
echo atmail.log.logins `date +%s` `echo "USE $MYSQL_DB; SELECT count(*) FROM Log_Login;" | mysql --host $MYSQL_HOST --user $MYSQL_USER -p$MYSQL_PASS --skip-column-names` host=$TAG_HOST
echo atmail.log.errors `date +%s` `echo "USE $MYSQL_DB; SELECT count(*) FROM Log_Error;" | mysql --host $MYSQL_HOST --user $MYSQL_USER -p$MYSQL_PASS --skip-column-names` host=$TAG_HOST

ENDTIME=$(date +%s%N)

echo tcollector.collectors.atmail.elapsedtime `date +%s` "$((($ENDTIME - $STARTTIME) / 1000))" host=$TAG_HOST


tc-opensips/tc-opensips.sh
#!/bin/sh
# 
# peritus-tc-tools/tc-opensips/tc-opensips.sh
# Count some tables from opensips MySQL database
# and output it to OpenTSDB tcollector
#
# 2014-04-06 Stian Ovrevage <stian@peritusconsulting.no>
#
# This script will simply count the number of rows for some of the opensips
# tables and output the metrics in tcollector compatible format.
#
# It will also output a metric of its own running time in microseconds in
# order to evaluate performance over time as counting mysql-rows on huge
# databases might result in performance degradation.
#
# Instructions:
#   Put this file into a tcollector/collectors/XX/ folder where XX is how
#   often you want the script to run. 60 should be reasonable and gives
#   statistics every minute. Make sure the script is runnable.  
# 
#   You can run the script on the opensips server itself or another host.
#   Make sure you create a user with permissions to connect if using a 
#   remote host. And update the config below.
#
#   You can run the script itself on the command-line to verify that 
#   it is working properly.
# 

MYSQL_USER=
MYSQL_PASS=
MYSQL_HOST=localhost
MYSQL_DB=opensips

TAG_HOST=`hostname`

STARTTIME=$(date +%s%N)

echo opensips.subscribers `date +%s` `echo "USE $MYSQL_DB; SELECT count(*) FROM subscriber;" | mysql mysql --host $MYSQL_HOST --user $MYSQL_USER -p$MYSQL_PASS --skip-column-names` host=$TAG_HOST
echo opensips.locations `date +%s` `echo "USE $MYSQL_DB; SELECT count(*) FROM location;" | mysql mysql --host $MYSQL_HOST --user $MYSQL_USER -p$MYSQL_PASS --skip-column-names` host=$TAG_HOST

ENDTIME=$(date +%s%N)

echo tcollector.collectors.opensips.elapsedtime `date +%s` "$((($ENDTIME - $STARTTIME) / 1000))" host=$TAG_HOST


tc-isc-dhcpd-log-stats/tc-isc-dhcpd-log-stats.sh.sh
#!/bin/sh
#
# peritus-tc-tools/tc-isc-dhcpd-log-stats.sh
# Count matching lines in ISC dhcpd log files
# and output it to OpenTSDB tcollector
#
# 2014-04-06 Stian Ovrevage <stian@peritusconsulting.no>
# 
# This script will simply grep words and sequences corresponding to
# dhcp-events from the ISC dhcpd log file.
#
# Instructions:
#   Put this file into a tcollector/collectors/XX/ folder where XX is how
#   often you want the script to run. 300 should be reasonable and gives
#   statistics every 5 minutes. Make sure the script is runnable.
#
# Requirements:
#   Be sure to enable logging in your dhcpd.conf:
#      log-facility local7
#   And log to a separate file in Xsyslog.conf:
#      local7.debug    /var/log/dhcpd.log
#   And to rotate the log file so that it will not fill your disks, or violate
#   customer privacy, /etc/logrotate.d/dhcpd:
#  
#      /var/log/dhcpd.log {
#          compress
#          dateext
#          maxage 365
#          daily
#          rotate 14
#          postrotate
#              /etc/init.d/rsyslog reload
#          endscript
#      }
# 

# instance tag sent with metrics, usefull if you have
# several instances, for example ipv4 and ipv6 
TAG_INSTANCE=ipv4

# dhcpd logfile location, default is /var/log/dhcpd.log
DHCPD_LOGFILE=/var/log/dhcpd.log

TAG_HOST=`hostname`


STARTTIME=$(date +%s%N)

echo dhcp.events.discover `date +%s` `grep DHCPDISCOVER $DHCPD_LOGFILE|wc -l` host=$TAG_HOST instance=$TAG_INSTANCE
echo dhcp.events.offer `date +%s` `grep DHCPOFFER $DHCPD_LOGFILE|wc -l` host=$TAG_HOST instance=$TAG_INSTANCE
echo dhcp.events.request `date +%s` `grep DHCPREQUEST $DHCPD_LOGFILE|wc -l` host=$TAG_HOST instance=$TAG_INSTANCE
echo dhcp.events.ack `date +%s` `grep DHCPACK $DHCPD_LOGFILE|wc -l` host=$TAG_HOST instance=$TAG_INSTANCE
echo dhcp.events.nak `date +%s` `grep DHCPNAK $DHCPD_LOGFILE|wc -l` host=$TAG_HOST instance=$TAG_INSTANCE
echo dhcp.events.release `date +%s` `grep DHCPRELEASE $DHCPD_LOGFILE|wc -l` host=$TAG_HOST instance=$TAG_INSTANCE
echo dhcp.events.inform `date +%s` `grep DHCPINFORM $DHCPD_LOGFILE|wc -l` host=$TAG_HOST instance=$TAG_INSTANCE

echo dhcp.events.addforwardmap `date +%s` `grep "Added new forward map from" $DHCPD_LOGFILE|wc -l` host=$TAG_HOST instance=$TAG_INSTANCE
echo dhcp.events.removeforwardmap `date +%s` `grep "Removed forward map from" $DHCPD_LOGFILE|wc -l` host=$TAG_HOST instance=$TAG_INSTANCE

echo dhcp.errors.noaddforwardmap `date +%s` `grep "Unable to add forward map from" $DHCPD_LOGFILE|wc -l` host=$TAG_HOST instance=$TAG_INSTANCE
echo dhcp.errors.noaddreversemap `date +%s` `grep "Unable to add reverse map from" $DHCPD_LOGFILE|wc -l` host=$TAG_HOST instance=$TAG_INSTANCE
echo dhcp.errors.notsec `date +%s` `grep "No tsec for use with key" $DHCPD_LOGFILE|wc -l` host=$TAG_HOST instance=$TAG_INSTANCE
echo dhcp.errors.bindupdatexidmismatch `date +%s` `grep "got ack from dhcp: xid mismatch" $DHCPD_LOGFILE|wc -l` host=$TAG_HOST instance=$TAG_INSTANCE
echo dhcp.errors.uidduplicate `date +%s` `grep "is duplicate on" $DHCPD_LOGFILE|wc -l` host=$TAG_HOST instance=$TAG_INSTANCE
echo dhcp.errors.bindupdaterejected `date +%s` `grep "incoming update is less critical than outgoing update" $DHCPD_LOGFILE|wc -l` host=$TAG_HOST instance=$TAG_INSTANCE

echo dhcp.errors.others `date +%s` `cat $DHCPD_LOGFILE | grep -v "DHCPDISCOVER" | grep -v "DHCPOFFER" | grep -v "DHCPREQUEST" | grep -v "DHCPACK" | grep -v "DHCPNAK" | grep -v "DHCPRELEASE" | grep -v "DHCPINFORM" | grep -v "Added new forward map from" | grep -v "Removed forward map from" | grep -v "Unable to add forward map from" | grep -v "Unable to add reverse map from" | grep -v "No tsec for use with key" | grep -v "got ack from dhcp: xid mismatch" | grep -v "is duplicate on" | grep -v "incoming update is less critical than outgoing update" | wc -l` host=$TAG_HOST instance=$TAG_INSTANCE

echo dhcp.total `date +%s` `cat $DHCPD_LOGFILE|wc -l` host=$TAG_HOST instance=$TAG_INSTANCE

ENDTIME=$(date +%s%N)

echo tcollector.collectors.isc-dhcpd-log-stats.elapsedtime `date +%s` "$((($ENDTIME - $STARTTIME) / 1000))" host=$TAG_HOST


tc-dhcpd-pools/tc-dhcpd-pools.py
#!/usr/bin/env python

"""

peritus-tc-tools/tc-dhcpd-pools/tc-dhcpd-pools.py
OpenTSDB tcollector wrapper for dhcpd-pools

2014-04-06 Stian Ovrevage <stian@peritusconsulting.no>

Runs dhcpd-pools and extracts data on dhcp pool usage and 
formats into OpenTSDB/tcollector compatible format.

Metrics gathered:
    dhcpd.pool.size - Total pool size, number of IPs
    dhcpd.pool.usage.absolute - Current pool usage, number of IPs (dhcpd-pools "current" field)
    dhcpd.pool.usage.percent - Current pool usage in percent (dhcpd-pools "percent" field)
    dhcpd.pool.inactive - Inactive, expired or abandoned leases (dhcpd-pools "touch" field)
    dhcpd.pool.backup - Failover backup allocatable addresses (dhcpd-pools "bu" field)

    For more information on the different fields, see dhcpd-pools documentation at:
        http://dhcpd-pools.sourceforge.net/man.html

dhcpd-pools output format:
  Shared network: 
    "name","max","cur","percent","touch","t+c","t+c perc","bu","bu perc"
    "prov-kvh018-RosRaadhus","234","16","6.838","23","39","16.667","0","0.000"

"""

import subprocess
import time
import sys
import os.path
import socket


''' dhcpd.conf file. Default is /etc/dhcpd.conf '''
config_file = '/etc/dhcpd.master'

''' dhcpd.leases file. Default is /var/db/dhcpd.leases '''
lease_file = '/var/db/dhcpd.leases'

''' host tag for metrics. Use system fqdn hostname by default '''
host = socket.getfqdn()


command = ['/usr/local/bin/dhcpd-pools --limit 02 --format c --config '
           + config_file + ' --leases ' + lease_file]


def main():
    if not os.path.isfile(config_file):
        print >> sys.stderr, 'error: config_file "'+config_file+'" does not exist'
        return 1
    if not os.path.isfile(lease_file):
        print >> sys.stderr, 'error: lease_file "'+lease_file+'" does not exist'
        return 1

    try:
        p = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, bufsize=0)
        line = p.stdout.readline()
        ts = int(time.time())               # Timestamp used when sending data
        while line:

            if len(line) > 1:                   # Only process lines with content

                line = line.replace('"','')     # Remove quotes
                fields = line.split(',')        # Split fields

                print("dhcpd.pool.size " + str(ts) + " " + fields[1]
                      + " host=" + host + " dhcppoolname=" + fields[0])
                print("dhcpd.pool.usage.absolute " + str(ts) + " " + fields[2]
                      + " host=" + host + " dhcppoolname=" + fields[0])
                print("dhcpd.pool.usage.percent " + str(ts) + " " + fields[3]
                      + " host=" + host + " dhcppoolname=" + fields[0])
                print("dhcpd.pool.inactive " + str(ts) + " " + fields[4]
                      + " host=" + host + " dhcppoolname=" + fields[0])
                print("dhcpd.pool.backup " + str(ts) + " " + fields[7]
                      + " host=" + host + " dhcppoolname=" + fields[0])

            line = p.stdout.readline()

    except OSError, err:
        print >> sys.stderr, 'Got execption running command "%s": %s' % (command, err)

if __name__ == "__main__":
    main()
